examples/dpo_demo_gemma3.ipynb

```diff
@@ -0,0 +1,849 @@
+   "outputs": [],
+   "source": [
+    "def extract_hash_answer(text: str) -> str | None:\n",
+    "  if \"####\" not in text:\n",
+    "    return None\n",
+    "  return text.split(\"####\")[1].strip()\n",
+    "\n",
+    "def get_dataset(data_dir, split=\"train\") -> grain.MapDataset:\n",
+    "  # Download data\n",
+    "  if not os.path.exists(data_dir):\n",
+    "    os.makedirs(data_dir)\n",
+    "\n",
+    "  data = tfds.data_source(\n",
+    "      \"gsm8k\",\n",
+    "      split=split,\n",
+    "      data_dir=data_dir,\n",
+    "      builder_kwargs={\"file_format\": tfds.core.FileFormat.ARRAY_RECORD},\n",
+    "      download=True,\n",
+    "  )\n",
+    "\n",
+    "  dataset = (\n",
+    "      grain.MapDataset.source(data)\n",
+    "      .shuffle(seed=42)\n",
+    "      .map(\n",
+    "          lambda x: {\n",
+    "              # passed to model forward pass\n",
+    "              \"prompts\": TEMPLATE.format(\n",
+    "                  system_prompt=SYSTEM_PROMPT,\n",
+    "                  question=x[\"question\"].decode(\"utf-8\"),\n",
+    "              ),\n",
+    "              # passed to reward functions\n",
+    "              \"question\": x[\"question\"].decode(\"utf-8\"),\n",
+    "              # passed to reward functions\n",
+    "              \"answer\": extract_hash_answer(x[\"answer\"].decode(\"utf-8\")),\n",
+    "          }\n",
+    "      )\n",
+    "  )\n",
+    "  return dataset\n",
+    "\n",
+    "test_dataset = get_dataset(TEST_DATA_DIR, \"test\").batch(BATCH_SIZE)[\n",
+    "    :NUM_TEST_BATCHES\n",
+    "]\n",
+    "\n",
+    "len(test_dataset)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "(corr, total, accuracy, partial_accuracy, format_accuracy), responses = evaluate(\n",
+    "    test_dataset,\n",
+    "    sampler,\n",
+    "    **GENERATION_CONFIGS[\"standard\"],\n",
+    "    make_lst=True\n",
+    ")\n",
+    "print(\n",
+    "    f\"{corr=}, {total=}, {accuracy=}%, {partial_accuracy=}%,\"\n",
+    "    f\" {format_accuracy=}%\"\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "LCAiINxUwYqx"
+   },
+   "source": [
+    "# Load DPO dataset"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "dpo_dataset = load_dataset(\"argilla/distilabel-intel-orca-dpo-pairs\", split=\"train\")\n",
+    "gsm8k_train_dpo_dataset = dpo_dataset.filter(lambda x: x['in_gsm8k_train'])\n",
+    "\n",
+    "# Get the number of samples in the filtered dataset\n",
+    "num_gsm8k_train_samples = len(gsm8k_train_dpo_dataset)\n",
+    "print(f\"Number of samples with in_gsm8k_train=True: {num_gsm8k_train_samples}\")\n",
+    "\n",
+    "# Calculate how many more samples are needed\n",
+    "total_samples_needed = NUM_BATCHES\n",
+    "samples_to_add = total_samples_needed - num_gsm8k_train_samples\n",
```

Comment:
Confused about this...this will work for batch size 1, but won't work for batch size > 1, right?

`gsm8k_train_dpo_dataset` is not batched above, maybe we should batch it up first?